README.md
# HMAC Auth Express

[![npm](https://img.shields.io/npm/v/hmac-auth-express)](https://www.npmjs.com/package/hmac-auth-express)
[![GitHub](https://img.shields.io/github/license/connorjburton/hmac-auth-express)](https://github.com/connorjburton/hmac-auth-express/blob/master/LICENSE)

This package provides [Express](https://expressjs.com/) middleware for [HMAC](https://en.wikipedia.org/wiki/HMAC) authentication.

- :star2: Zero dependencies
- :clock1: Timing safe
- :white_check_mark: 100% code coverage
- :books: Support for all hash algorithms
- :lock: Replay attacks prevention

## Installation

**npm**

`npm install hmac-auth-express`

**yarn**

`yarn add hmac-auth-express`

## Type documentation

[View the documentation online here](https://connorjburton.github.io/hmac-auth-express), or run `npm run docs` in the repository.

## Usage

#### Importing the package

Supports both CommonJS & ECMAScript modules.

_CommonJS_

```javascript
const { HMAC } = require("hmac-auth-express");
```

_ECMAScript_

```javascript
import { HMAC } from "hmac-auth-express";
```

See [FAQs](#faqs) for design decisions around exports.

#### Basic middleware registration

```javascript
app.use("/api", HMAC("secret"));
```

#### Advanced middleware registration

```javascript
app.use(
  "/api",
  HMAC("secret", {
    algorithm: "sha512",
    identifier: "APP",
    header: "myheader",
    maxInterval: 600,
    minInterval: 600,
  }),
);
```

[See the docs for full list of options.](https://connorjburton.github.io/hmac-auth-express/interfaces/Options.html)

If you are using this package to authenticate against routes with bodies, you must register the middleware after you have parsed and set the `request.body` property. The simplest way to achieve this is to use the built-in `express.json()` method.

```javascript
app.use(express.json());
app.use("/api", HMAC("secret"));
```

#### Error Handling

The middleware will throw `TypeError`'s if you provide incorrect parameters at registration time.

The middleware will pass an error to [express' error handler](http://expressjs.com/en/guide/error-handling.html#writing-error-handlers). You can use the provided `AuthError`, or alternatively check the error by its code `ERR_HMAC_AUTH_INVALID`.

Example:

```javascript
const { AuthError } = require("hmac-auth-express");

// express' error handler
app.use((error, req, res, next) => {
  // check by error instance
  if (error instanceof AuthError) {
    res.status(401).json({
      error: "Invalid request",
      info: error.message,
    });
  }

  // alternative: check by error code
  if (error.code === "ERR_HMAC_AUTH_INVALID") {
    res.status(401).json({
      error: "Invalid request",
      info: error.message,
    });
  }
});
```

## Dynamic Secret

From [8.2.0](https://github.com/connorjburton/hmac-auth-express/releases/tag/v8.2.0) onwards you can alternatively supply a function as your `secret` parameter. This function accepts 1 parameter, being an `express.Request` object. This function can be `async`. You can use this feature to dynamically determine the secret of the request, for example if you have different HMAC secrets depending on the URL of the request.

```javascript
const dynamicSecret = async (req) => {
  // determine and return what the secret should be from the request object

  return myDynamicSecret;
};

app.use(HMAC(dynamicSecret));
```

## Structuring your HMAC header

Now you have configured your HMAC middleware, you need to structure your HMAC in the way the middleware expects.

#### What your HMAC should look like

_This example uses the default `options.header` and `options.identifier`._

`Authorization: HMAC 1573504737300:76251c6323fbf6355f23816a4c2e12edfd10672517104763ab1b10f078277f86`

#### Constructing the HMAC

`Authorization:` This is the header you send in the request that contains the HMAC. This is what the middleware will look for.

`HMAC` This is the identifier the middleware will look for.

`1573504737300` This is the UNIX timestamp of when the request was sent.

`76251c6323fbf6355f23816a4c2e12edfd10672517104763ab1b10f078277f86` This is the HMAC digest, see [generating your HMAC digest](#generating-your-hmac-digest).

## Generating your HMAC digest

The HMAC signature is 4 parts (1 part optional) joined **without** a seperator.

| Part           | Required | Example                            | Description                                   |
| -------------- | -------- | ---------------------------------- | --------------------------------------------- |
| Unix Timestamp | Yes      | `1573504737300`                    | The current unix timestamp                    |
| Verb           | Yes      | `POST`                             | The verb of the request you are making        |
| Route          | Yes      | `/api/order`                       | The route you are requesting                  |
| MD5 JSON Body  | No       | `9bb58f26192e4ba00f01e2e7b136bbd8` | The MD5 of your `JSON.stringify` request body |

Below is an example request and how we would build that requests HMAC

```
POST http://www.domain.com/api/order HTTP/1.0
Authorization: HMAC 1573504737300:76251c6323fbf6355f23816a4c2e12edfd10672517104763ab1b10f078277f86
Content-Type: application/json
Date: Wed, 13 Nov 2019 22:06:01 GMT

{
  "foo": "bar"
}
```

```javascript
const crypto = require("crypto");

const hmac = crypto.createHmac("sha256", "secret");
const time = Date.now().toString();

hmac.update(time);
hmac.update("POST");
hmac.update("/api/order");

const contentHash = crypto.createHash("md5");
contentHash.update(JSON.stringify({ foo: "bar" }));

hmac.update(contentHash.digest("hex"));

console.log(`HMAC ${time}:${hmac.digest("hex")}`);
```

You can also use the [exported generate function](https://connorjburton.github.io/hmac-auth-express/modules.html#generate) if you are using JavaScript on your client.

```javascript
const { generate } = require("hmac-auth-express");

const time = Date.now().toString();
const digest = generate("secret", "sha256", time, "POST", "/api/order", {
  foo: "bar",
}).digest("hex"); // 76251c6323fbf6355f23816a4c2e12edfd10672517104763ab1b10f078277f86

const hmac = `HMAC ${time}:${digest}`;
```

## Replay attacks

The parameter `options.maxInterval` is the amount of time in seconds that a request is valid. We compare the unix timestamp sent in the HMAC header to the current time on the server. If the time difference is greater than `options.maxInterval` request is rejected.

Similarly `options.minInterval` (introduced in [4.1.0](https://github.com/connorjburton/hmac-auth-express/releases/tag/v4.1.0)) is the amount of time in seconds that a request is valid for if in the future. This is a common issue for out of sync computer times (the requester time is slightly ahead of the server). If you find requests being rejected as they are from the future you may want to adjust this.

The unix timestamp sent in the header is also included in the HMAC digest, this is to prevent someone replicating a request and changing the unix timestamp to be in valid range of `options.maxInterval` or `options.minInterval`

## How to handle non-deterministic JSON

In Node.js, JSON parsing is non-deterministic. This poses problems when depending on that order to build HMAC digests, as a difference in order of the request body between client and server will result in different HMAC digests and fail.

There are various solutions to this problem, with various levels of effort required. As this middleware intends to abstract away complexity in how to handle HMAC we have provided helpers and extensibility to help you solve this problem. This is available in version [8.3.0](https://github.com/connorjburton/hmac-auth-express/releases/tag/v8.3.0).

If you require your JSON parsing to be deterministic, you can use the [exported `order` function](https://connorjburton.github.io/hmac-auth-express/modules.html#order) and pass that as an [argument to `HMAC`](https://connorjburton.github.io/hmac-auth-express/interfaces/Options.html#order).

```javascript
import { HMAC, order } from "hmac-auth-express";

app.use(HMAC("secret", { order }));
```

You will need to do the same on your client, the [`generate` method](https://connorjburton.github.io/hmac-auth-express/modules.html#generate) accepts an `options.order` parameter.

```javascript
const { generate, order } = require("hmac-auth-express");

const digest = generate(
  "secret",
  "sha256",
  Date.now().toString(),
  "POST",
  "/api/order",
  { foo: "bar" },
  { order },
).digest("hex"); // 76251c6323fbf6355f23816a4c2e12edfd10672517104763ab1b10f078277f86

const hmac = `HMAC ${Date.now().toString()}:${digest}`;
```

This exported function recursively orders your object in lexigraphic order. This should be sufficient in most cases.

You can provide your own order function with your own implementation if you wish.

## Limitations

This package does not support plain text, form or multi part POST bodies and is primarily intended to be used for GET requests and JSON bodies. [Plain text support](https://github.com/connorjburton/hmac-auth-express/issues/61) is planned.

Be mindful of what algorithm you choose to use, this package will not stop you attempting to use an algorithm that is not supported by OpenSSL. [See the Node.js website for more information](https://nodejs.org/en/knowledge/cryptography/how-to-use-crypto-module/#hash-algorithms-that-work-with-crypto).

## Performance

You can run your own benchmarks by cloning the repository and running `npm run benchmark`. Below is an example result, however please run the benchmarks on your intended target machine for accurate results. Node.js >16 is required to run the benchmark tool.

| Environment                        | Operations | Duration | ops/second |
| ---------------------------------- | ---------- | -------- | ---------- |
| `Windows 10 Pro, i5-7600K@3.80GHz` | 1,000,000  | 752ms    | 1,329,767  |

## Deno support

This package exports two main components, an Express middleware and a HMAC generator. The HMAC generator does work and is supported in Deno. The Express middleware would likely also work with Deno, however I am not aware of a Deno router that has the same function signature for middleware as Express.

An example of how to use this package in Deno

```javascript
import { generate } from "npm:hmac-auth-express@8.3.2";

const digest = generate(
  "secret",
  "sha256",
  Date.now().toString(),
  "POST",
  "/api/order",
  { foo: "bar" },
).digest("hex");

const hmac = `HMAC ${Date.now().toString()}:${digest}`;
console.log(`HMAC is ${hmac}`);
```

In the future we may expose the core middleware logic standalone, so that you can inject this yourself in any middleware function signature of your choosing.

## Tests

100% code coveage is a hard requirement of this package, enforced with automated coverage checks. The [GitHub action](https://github.com/connorjburton/hmac-auth-express/blob/master/.github/workflows/ci.yml) will use `awk` to ensure that the action fails if any coverage criteria is less than 100% for any source file.

Running the tests requires Node.js v20 or higher.

To run tests run `npm run test`

To run tests with coverage run `npm run test -- --experimental-test-coverage`

## FAQs

_Why is HMAC uppercase?_ HMAC is an acronym for [hash-based message authentication code](https://en.wikipedia.org/wiki/HMAC). You can import the package as below if you need to conform to style conventions.

```javascript
import { HMAC as hmac } from "hmac-auth-express";
```

_Why is there no default export?_ It seems to be non-trivial to export a default that has consistent behaviour between CommonJS & ECMAScript, the example below shows the behavioural differences when exporting a default from TypeScript.

```javascript
const HMAC = require("hmac-auth-express").default;
import HMAC from "hmac-auth-express";
```

If you have a suggestion on how to export a default consistently then please [open an issue](https://github.com/connorjburton/hmac-auth-express/issues/new).

_Why is MD5 used instead of x?_ We use MD5 to create a hash of the request body (if available) as part of building the digest, which is then hashed in totality with SHA256 (by default), therefore we are not using MD5 to secure any part of this package. MD5 is used as it is faster than the built-in alternatives and adding an external dependency for a faster hashing algorithm such as Murmur is unnecessary.

## Credits

Reference article https://www.wolfe.id.au/2012/10/20/what-is-hmac-authentication-and-why-is-it-useful/


CHANGELOG.md
# [9.0.0](https://github.com/connorjburton/hmac-auth-express/compare/v8.3.5...v9.0.0) (2023-09-17)


* fix!: order function correctly orders arrays ([172598d](https://github.com/connorjburton/hmac-auth-express/commit/172598d0916eb6db8eacaf626184d50bdc7b2b7d))


### BREAKING CHANGES

* the order function was not ordering array values
which would lead to non-deterministic ordering if using arrays

## [8.3.5](https://github.com/connorjburton/hmac-auth-express/compare/v8.3.4...v8.3.5) (2023-09-17)


### Bug Fixes

* use file extension to meet ts requirementnts ([5e867cb](https://github.com/connorjburton/hmac-auth-express/commit/5e867cb110d8021a7b55065e006543f19d44bc1a))


### Reverts

* Revert "chore: debug gh action for exit code" ([f59439d](https://github.com/connorjburton/hmac-auth-express/commit/f59439d35e8fadc3721655cbf178cb3938654ef4))

## [8.3.4](https://github.com/connorjburton/hmac-auth-express/compare/v8.3.3...v8.3.4) (2023-02-14)


### Bug Fixes

* remove badge as reporting incorrect statuses ([a069683](https://github.com/connorjburton/hmac-auth-express/commit/a06968397f47691252bf65d246741f5faee5f28c))

## [8.3.3](https://github.com/connorjburton/hmac-auth-express/compare/v8.3.2...v8.3.3) (2023-02-13)

### Refactor

* Remove `stringToBuffer` and replace with `Buffer#from` ([2597079][https://github.com/connorjburton/hmac-auth-express/commit/2597079f5f0ed519fa439460454ab3a69486d353])

## [8.3.2](https://github.com/connorjburton/hmac-auth-express/compare/v8.3.1...v8.3.2) (2022-03-13)

### Bug Fixes

* reduce published package size from 334kb to 36kb ([e7d5801](https://github.com/connorjburton/hmac-auth-express/commit/e7d5801c4ad76c79f27461587c5122dd87f4bf2e))

## [8.3.1](https://github.com/connorjburton/hmac-auth-express/compare/v8.3.0...v8.3.1) (2022-03-13)

### Bug Fixes

* remove a === b in order func as not a possible state ([5bca1e4](https://github.com/connorjburton/hmac-auth-express/commit/5bca1e4e78ef09407a090224ef5d434c706a12a8))

## [8.3.0](https://github.com/connorjburton/hmac-auth-express/compare/v8.2.0...v8.3.0) (2022-03-07)

### Features

* add option to order body ([c7558d1](https://github.com/connorjburton/hmac-auth-express/commit/c7558d1d7c8d74f19b8e33128073d293a48ee94c))


## [8.2.0](https://github.com/connorjburton/hmac-auth-express/compare/v8.1.0...v8.2.0) (2021-08-23)


### Features

* allow secret parameter to be a function to support dynamic secrets ([e612eff](https://github.com/connorjburton/hmac-auth-express/commit/e612effcf7adea2541d33593880033e14f9be0a9))

## [8.1.0](https://github.com/connorjburton/hmac-auth-express/compare/v8.0.2...v8.1.0) (2021-07-18)


### Features

* exposing generate function to easily generate hmac digests ([7996b40](https://github.com/connorjburton/hmac-auth-express/commit/7996b40d1ba958d63546e3d619ebba65a0a8d837))


### Bug Fixes

* updating benchmark tool to use new default header ([7836d75](https://github.com/connorjburton/hmac-auth-express/commit/7836d7547f1c09e02488feb2c1be652cef42ef07))

## [8.0.1-8.0.2](https://github.com/connorjburton/hmac-auth-express/compare/v8.0.0...v8.0.2) (2021-07-11)

### Bug Fixes

* published version did not include all compiled files correctly

## [8.0.0](https://github.com/connorjburton/hmac-auth-express/compare/v7.0.0...v8.0.0) (2021-07-11)


### [⚠ BREAKING CHANGES](https://github.com/connorjburton/hmac-auth-express/blob/master/MIGRATION_GUIDE.md)

* changed default header to authorization
* package now exclusively uses named exports
* error is now exported as `AuthError` rather than `HMACAuthError`
* importing `AuthError` is now done directly on `hmac-auth-express` rather than `hmac-auth-express/src/errors`

### Bug Fixes

* now supports empty objects and arrays as request bodies ([640bcbe](https://github.com/connorjburton/hmac-auth-express/commit/640bcbe323897b1100961c226078385f83a3121e))

## [7.0.0](https://github.com/connorjburton/hmac-auth-express/compare/v6.0.1...v7.0.0) (2021-07-04)


### ⚠ BREAKING CHANGES

* addition of express peer dependency may fail installations of users using express <4

### Bug Fixes

* changes to package.json & e2e test ([4b17f33](https://github.com/connorjburton/hmac-auth-express/commit/4b17f33d4107a47453237781eeccb845347159d0))

## 6.0.1 - 2021-06-28
### Added
- Added Github Actions
### Removed
- Removed Travis integration
### Security
- Bumped `jest` from `26.3.0` to `27.0.6`

## 6.0.0 - 2020-08-12
### Changed
- **Breaking** Inversed `timeDiff` value in comparison against `minInterval` option so option parameters are consistent.

## 5.0.1 - 2020-08-12
### Changed
- Updated `parseInt` call to pass `radix` parameter of `10`
### Security
- Bumped `jest` from `26.2.2` to `26.3.0`

## 5.0.0 - 2020-08-01
### Added
- **Breaking** Support for plain arrays in the request body. Previously a plain array request body would be rejected, this change could potentially allow unexpected behaviour, thus the breaking change.
### Security
- Bumped `jest` from `24.9.0` to `26.2.2`

## 4.1.0 - 2020-03-20
### Added
- `options.minInterval` for out of sync times (requests from the future)
### Security
- Bumped `acorn` from `5.7.3` to `5.7.4`

## 4.0.2 - 2019-12-31
### Added
- Added Travis CI for running tests on pull requests
- Added automatic NPM publish on master push
### Changed
- Added Travis CI badge to readme

## 4.0.1 - 2019-11-13
### Changed
- Updated changelog to include correct date for 4.0.0

## 4.0.0 - 2019-11-13
### Added
- Tests, 100% coverage
- `.vscode/launch.json` that includes test configuration for easily debugging tests

### Changed
- **Breaking** The UNIX timestamp's used is now expected to be 13 characters long, not 10 (i.e already divided by `1000`). See [Migration Guide](https://github.com/connorjburton/hmac-auth-express/blob/master/MIGRATION_GUIDE.md#migrating-from-3xx-to-400)
- Improved validation for checking if the HMAC digest exists in the header
- Improved validation for checking if `request.body` is set

## 3.0.0 - 2019-01-02
### Changed

src/index.ts
import { Request, RequestHandler, NextFunction } from "express";

import validateArguments from "./validateArguments.js";
import AuthError from "./errors.js";
import generate, { GenerateOptions } from "./generate.js";
import order from "./order.js";

const crypto = await import("node:crypto");

export { AuthError, generate, GenerateOptions, order };

export type UnknownObject = Record<string, unknown>;
/**
 * Your hash secret or [a function](https://github.com/connorjburton/hmac-auth-express#dynamic-secret)
 *
 * Defaults to `sha256`
 */
export type DynamicSecret =
  | string
  | ((req: Request) => string | undefined)
  | ((req: Request) => Promise<string | undefined>);
export type Order = (o: UnknownObject | unknown[]) => UnknownObject | unknown[];

export interface Options {
  /**
   * The hashing algorithm
   *
   * Defaults to `sha256`
   */
  algorithm: string;
  /**
   * The start of your `options.header` should start with this
   *
   * Defaults to `HMAC`
   */
  identifier: string;
  /**
   * The header the HMAC is located
   *
   * Defaults to `authorization`
   */
  header: string;
  /**
   * The amount of time you would like a request to be valid for,
   * in seconds (in the past).
   * See [time based protection against replay attacks](https://github.com/connorjburton/hmac-auth-express#replay-attacks) for more information
   *
   * Defaults to `300`
   */
  maxInterval: number;
  /**
   * The amount of time you would like a request to be valid for,
   * in seconds (in the future).
   * See [time based protection against replay attacks](https://github.com/connorjburton/hmac-auth-express#replay-attacks) for more information
   *
   * Defaults to `0`
   */
  minInterval: number;
  /**
   * Optional function to order your object before stringifying, see [How to handle non-deterministic JSON](https://github.com/connorjburton/hmac-auth-express#how-to-handle-non-deterministic-json)
   */
  order?: Order;
}

export const defaults: Options = {
  algorithm: "sha256",
  identifier: "HMAC",
  header: "authorization",
  maxInterval: 60 * 5,
  minInterval: 0,
};

async function determineSecret(
  secret: DynamicSecret,
  req: Request,
): Promise<string | undefined> {
  return typeof secret === "function" ? await secret(req) : secret;
}

export function HMAC(
  secret: DynamicSecret,
  options: Partial<Options> = {},
): RequestHandler {
  const mergedOpts: Options = { ...defaults, ...options };

  validateArguments(secret, mergedOpts);

  return async function (
    request: Request,
    _,
    next: NextFunction,
  ): Promise<void> {
    // we have to create a scoped secret per request, if we were to reassign the original `secret` variable
    // the next request that comes in will no longer be the secret function
    // this means we have to explicitly check it's not undefined too
    const scopedSecret = await determineSecret(secret, request);
    if (typeof scopedSecret !== "string" || scopedSecret.length === 0) {
      return next(
        new AuthError(
          `Invalid secret. Expected non-empty string but got '${scopedSecret}' (type: ${typeof scopedSecret})`,
        ),
      );
    }

    const header = request.get(mergedOpts.header);
    if (typeof header !== "string") {
      return next(
        new AuthError(
          `Header provided not in sent headers. Expected ${mergedOpts.header} but not found in request.headers`,
        ),
      );
    }

    if (
      typeof mergedOpts.identifier !== "string" ||
      !header.startsWith(mergedOpts.identifier)
    ) {
      return next(
        new AuthError(
          `Header did not start with correct identifier. Expected ${mergedOpts.identifier} but not found in options.header`,
        ),
      );
    }

    const unixRegex = /(\d{1,13}):/;
    const unixMatch = unixRegex.exec(header);
    if (
      !Array.isArray(unixMatch) ||
      unixMatch.length !== 2 ||
      typeof unixMatch[1] !== "string"
    ) {
      return next(new AuthError("Unix timestamp was not present in header"));
    }

    // is the unix timestamp difference to current timestamp larger than maxInterval or smaller than minInterval
    const timeDiff =
      Math.floor(Date.now() / 1000) -
      Math.floor(parseInt(unixMatch[1], 10) / 1000);
    if (
      timeDiff > mergedOpts.maxInterval ||
      timeDiff * -1 > mergedOpts.minInterval
    ) {
      return next(
        new AuthError(
          "Time difference between generated and requested time is too great",
        ),
      );
    }

    // check HMAC digest exists in header
    const hashRegex = /:(.{1,}$)/;
    const hashMatch = hashRegex.exec(header);
    if (
      !Array.isArray(hashMatch) ||
      hashMatch.length !== 2 ||
      typeof hashMatch[1] !== "string"
    ) {
      return next(new AuthError("HMAC digest was not present in header"));
    }

    const hmac = generate(
      scopedSecret,
      mergedOpts.algorithm,
      unixMatch[1],
      request.method,
      request.originalUrl,
      request.body,
      { order: mergedOpts.order },
    ).digest();
    const sourceDigest = Buffer.from(hashMatch[1], "hex");

    // use timing safe check to prevent timing attacks
    if (
      hmac.length !== sourceDigest.length ||
      !crypto.timingSafeEqual(hmac, sourceDigest)
    ) {
      return next(new AuthError("HMAC's did not match"));
    }

    return next();
  };
}
